### ISSUE TITLE
C++ accessors for value instances are not "const".

### ISSUE THREAD

**Issue by webbnh** [kaitai-io/kaitai_struct#294]
The accessor functions for type fields are marked `const`, e.g.:
```
    public:
        double val_in_rads() const { return m_val_in_rads; }
```
but the accessor for the related instance is not:
```
    public:
        double val_in_degrees();
```
```
double mystruct_t::loc_status_t::val_in_degrees() {
    if (f_val_in_degrees)
        return m_val_in_degrees;
    m_val_in_degrees = ((val_in_rads() * 180.0) / 3.14159265);
    f_val_in_degrees = true;
    return m_val_in_degrees;
}
```
Is this intentional?

It's annoying, because my higher level code doesn't distinguish between the two types of accessor functions (and, because the higher level code receives a pointer-to-const parameter, so it can't call the instance accessor function without casting away the `const`, which is _ugly_).

Thanks!

**Comment by GreyCat** [comment]
Good point! No, it's not intentional, and, actually, instances should be exactly read-only by design.

**Comment by webbnh** [comment]
Thanks.

**Comment by GreyCat** [comment]
Revisiting this issue, I've become somewhat hesitating about declaring everything `const`. I'm totally not a big guru of C++, but, technically, `const` on a non-static method means that this method is not changing anything in object's internal state. And with instances, this is exactly what's happening: first invocation of this method will do the reading and fill in some fields (and actually reading might even throw an exception). Subsequent invocations of the same methods should be exception-safe and indeed do no changes to object's internal state.

Does that really qualify for being `const`? I've checked at least gcc7, and it seems to consider that an error.

```cpp
std::string instance_std_t::header() const {
    if (f_header)
        return m_header;
    std::streampos _pos = m__io->pos();
    m__io->seek(2);
    m_header = kaitai::kstream::bytes_to_str(m__io->read_bytes(5), std::string("ASCII"));
    m__io->seek(_pos);
    f_header = true;
    return m_header;
}
```

yields

    instance_std.cpp:27:88: error: passing ‘const string {aka const std::__cxx11::basic_string<char>}’ as ‘this’ argument discards qualifiers [-fpermissive]
    instance_std.cpp:29:16: error: assignment of member ‘instance_std_t::f_header’ in read-only object

It turns out that it would be probably impossible :(

**Comment by webbnh** [comment]
As an API-provider, I _hate_ `const`; however, as an API-consumer, I very much appreciate having it available.

Conceptually, instance accessor methods _should_ be `const`, because, semantically, they don't modify the object (e.g., they are directly analogous to field accessors).  Of course, as you point out, under the covers they are setting up and caching values, so they need to modify some fields in the object.

If you're willing to journey down the rabbit hole (I _did_ say that, in your shoes, I _hate_ `const`... ;-) ), I think you're looking for the [mutable keyword](http://en.cppreference.com/w/cpp/language/cv) ([here's](https://www.geeksforgeeks.org/c-mutable-keyword/) a nice example explanation) which would allow you to mark the accessor methods as `const` but still allow their implementations to do things like cache values (by marking those fields as `mutable`).

**Comment by arekbulski** [comment]
C++ syntax is a sin in its own right. :expressionless: 

@webbnh 
If you dont mind me asking, what is the project where you use Kaitai? And does it use C++ runtime? I am just being curious here.

**Comment by GreyCat** [comment]
@webbnh Thanks for the tip! Indeed, `mutable` is the solution. I'm slightly hesitating about it, as I can't seem to find which C++ version it was introduced to. Our current commitment is to support of very old C++ standards, i.e. almost down to C++98. Any idea if `mutable` might work in ancient compilers?

**Comment by webbnh** [comment]
@arekbulski, we use KS to parse embedded device log files.  We're currently rewriting our Python prototype in C++ (for a number of reasons).  We describe the various log file formats in a bunch of `.ksy` files, we use `ksc` to generate the code to read them, we use a tool I wrote to generate code from the `.ksy` files to produce JSON output, and, yes, we use the KS C++ runtime to make it all work.

@GreyCat, `mutable` is very old in C++ terms.  Obviously, it's no older than `const`, but I'm pretty sure that the need for it became apparent shortly after `const` was proposed, and they probably came out in the same version.

As for being hesitant to use it, that's healthy.  :-)  Using `mutable` requires discipline and a certain amount of attention.  Using `const` is highly inconvenient and imposes all sorts of hassle, but the results are a very comfortable degree of safety and reliability.  Using `mutable` pokes holes in that safety and reliability, and, once you start using it, the temptation to expand uses of it to relieve inconvenience and to deal with hassle becomes a significant force.  Generally, the best approach is to avoid that temptation by not using it.  However, in this case, I think using `mutable` is the right thing to do, but you'll need to be vigilant once it's in the code.  :-)

**Comment by Wyverex42** [comment]
Is this issue still active? I'm currently using Kaitai in a C++ context (it's a great help btw!) and I have to do some awkward hacks to get around the fact that accessors are not `const` as they should be. `const` correctness is a big deal in terms of API design and should be observed wherever possible. See also related Iso C++ guidelines (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#con2-by-default-make-member-functions-const).

Using `mutable` for instances is the correct solution. The fact that those values (and the associated `bool`) might not have been calculated yet when the function is called is irrelevant from the caller's point of view as long as the value doesn't change between function calls (which doesn't happen as far as I can see in the generated code). For the caller it's just a simple getter access which should be `const` if possible to preserve compatibility with other APIs.

I appreciate that it's a bit of work to add this in hindsight, but fortunately it's not challenging, it's just a PITA. Do you have any plans to address this?

Thanks!
